pkg/template/params.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package template
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	pipelinev1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// AddDefaultParamsFromSpec returns the params with the addition of all
+// paramSpecs that have default values and are already in the params list
+func AddDefaultParamsFromSpecs(params []pipelinev1.Param, paramSpecs []pipelinev1.ParamSpec) []pipelinev1.Param {
+	allParamsMap := map[string]pipelinev1.ArrayOrString{}
+	for _, paramSpec := range paramSpecs {
+		if paramSpec.Default != nil {
+			allParamsMap[paramSpec.Name] = *paramSpec.Default
+		}
+	}
+	for _, param := range params {
+		allParamsMap[param.Name] = param.Value
+	}
+	allParams := []pipelinev1.Param{}
```

Comment:
Because we know that the size of `allParams` will invariably be the same size as `allParamsMap`, if we initialize the slice with that known size the slice won't have to repeatedly resized as we append elements; I think this line would be better: `allParams := make([]pipelinev1.Param, len(allParamsMap))`

Reply:
Thanks, good point 👍